[Skeltons] iP

Skeltons opened the pull request:
# Duke #
> The most important step a man can take. It's not the first one, is it? It's the next one. - Brandon Sanderson
>
Duke frees your mind of having to remember tasks! It's,
* fast
* text based
* and friendly!

All you need to do is,
1. Download it from [here](https://github.com/Skeltons/ip/releases/tag/A-Jar)
2. Navigate to where you downloaded Duke
3. Use your terminal to run it with `java -jar Duke.jar`
4. Start using it! :smile:

And it is **FREE!**

Features:
- [x] Manage tasks ~~(coming soon)~~
- [ ] Manage deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practise Java too. Here's the `main` method:

```java
public static void main(String[] args) {
    new Duke("duke").chat();
}
```

Commit:
Add Gradle support

Skeltons committed:
Level-1: Greet, Echo, Exit

Skeltons committed:
Level 2. Add, List

Skeltons committed:
Level-3 Mark as Done

Skeltons committed:
Modified .gitignore

Skeltons committed:
Level-4: ToDo, Event, Deadline

Skeltons committed:
A-TextUiTesting: Automated Text UI Testing

Skeltons committed:
Level-5: Handle Error

Skeltons committed:
Level 6: Delete

Skeltons committed:
Fix indentation on switch-case

Skeltons committed:
Add function that saves tasklist after every change

Skeltons committed:
Add LocalDateTime attribute to deadline class

Skeltons committed:
Resolve merge conflict with branch-Level-8

Skeltons committed:
Merge branch 'branch-Level-8'

Skeltons committed:
Add duke tasklist (duke.txt) to .gitignore

Skeltons committed:
Make the code more OOP

Skeltons committed:
Add source file to their respective package

Skeltons committed:
Add java files to their respective package

Skeltons committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Skeltons committed:
Set up gradle

Skeltons committed:
Set up gradle

Skeltons committed:
Modify gradle settings

Skeltons committed:
Add JUnit tests to test ToDoCommand and ByeCommand

Skeltons committed:
Change class name in build.gradle

Skeltons committed:
Add JavaDoc to public classes and methods

Skeltons committed:
Remove .project and .classpath from repo

Skeltons committed:
Add function to find task based on keywords

Skeltons committed:
Add .project into .gitignore

Skeltons committed:
Add .project into .gitignore

Skeltons committed:
Update gitignore

Skeltons committed:
Refactor code to Java coding standard.

Skeltons committed:
Merge branch 'branch-A-JavaDoc'

Skeltons committed:
Merge master with branch-A-JavaDoc

Skeltons committed:
Remove arguement from TaskList

Skeltons committed:
Merge branch 'branch-A-JavaDoc'

Skeltons committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/ui/Ui.java

Skeltons committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/ui/Ui.java

Skeltons committed:
Resolve merge conflict between master and branch-Level-9

Skeltons committed:
Fix bug with mark and unmark command

Skeltons committed:
Refactor code to coding standards after checkstyle

Skeltons committed:
Rename boolean variable to match coding standard

Skeltons committed:
Rename boolean variables to match coding standards

Skeltons committed:
Add GUI to Duke with JavaFX

Skeltons committed:
Change main class to Launcher for shadowJar

Oliverloo0909 commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    public Command parse(String echo) {
+        String[] command = echo.split(" ");
+        switch (command[0]) {
```
I like how you used the switch statement to make all commands clear and concise :)

Oliverloo0909 commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,75 @@
+            StringBuilder endTime = new StringBuilder();
+            boolean hasFrom = false;
+            boolean hasTo = false;
+            for (int i = 1; i < command.length; i++) {
```
Is there a more efficient way to do this? Feels Like a rather long and complicated method, just to segment the "/to" and "/from" strings. I understand you are working with an array, but perhaps rejoining the command string form [1:end] and using the String.split() method could make your life easier?

Oliverloo0909 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+ * @version 0.1
+ * @since 0.1
+ */
+public class Task {
```
Since there is no instance of a task command. Should this class be abstract?

Oliverloo0909 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,56 @@
+ * @since 0.1
+ */
+public class TaskList {
+    private List<Task> taskList = new LinkedList<>();
```
Interesting! Is there a reason why you a LinkedList instead of an ArrayList?

homuntan02 replied:
I agree very interesting indeed.

Oliverloo0909 reviewed (commented):
Overall, very well-written code :) Masterful use of OOP, everything was well encapsulated. Adhered to the coding standard, java docs were easy understand. Fantastic job!

homuntan02 commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,21 @@
+    public DukeException(String error, Throwable err) {
+        super(error, err);
+    }
+
```
Maybe you can try add more types of exceptions to be more specific.

homuntan02 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,54 @@
+public class Event extends Task {
+    protected String startTime;
+    protected String endTime;
+    protected HashSet<String> wordsInDescription = new HashSet<>();
```
Interesting use of hashset to quicken find. Nice.

homuntan02 commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,49 @@
+package duke.command;
```
Maybe try putting parser into it's own package since it's not a command.

homuntan02 commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,49 @@
+            return new DeleteCommand(command);
+        case "find":
+            return new FindCommand(command);
+        default:
```
good use of switchcases.

Skeltons committed:
Fix bug where storage was not loading stored data

Skeltons committed:
Fix bug where event task is not being loaded from storage.

Skeltons committed:
Fix bug with unmark and mark not loading.

Skeltons committed:
Add assertions to tasks

Assertions is added to ensure that the description
of tasks is not an empty String.

This is important because a task cannot have an
empty description, thus making this assertion
on its description important.

Skeltons committed:
Merge pull request #1 from Skeltons/branch-A-Assertions

Add assertions to tasks

Skeltons committed:
Improve Code Quality

Skeltons committed:
Merge pull request #2 from Skeltons/branch-A-CodeQuality

Improve Code Quality

Skeltons committed:
Add the ability to tag tasks.

Skeltons committed:
Refactor code such that it aligns with coding style

Skeltons committed:
Change duke and user images to avoid copyright

Skeltons committed:
Modify DialogBox.fxml to not allow text overflow

Skeltons committed:
Give credit for reused work

Skeltons committed:
Update Userguide for Duke Bunny

Skeltons committed:
Fix indentation of docs/README.md

Skeltons committed:
Fix indentation in UG
